src/hotspot/share/gc/shared/weakProcessor.inline.hpp

```diff
@@ -81,13 +81,12 @@ void WeakProcessor::Task::work(uint worker_id,
          "worker_id (%u) exceeds task's configured workers (%u)",
          worker_id, _nworkers);
 
-  typedef WeakProcessorPhases::Iterator Iterator;
-
-  for (Iterator it = WeakProcessorPhases::oopstorage_iterator(); !it.is_end(); ++it) {
-    WeakProcessorPhase phase = *it;
+  constexpr EnumRange<WeakProcessorPhase> phase_range{};
+  for (WeakProcessorPhase phase : phase_range) {
     CountingClosure<IsAlive, KeepAlive> cl(is_alive, keep_alive);
     WeakProcessorPhaseTimeTracker pt(_phase_times, phase, worker_id);
-    StorageState* cur_state = _storage_states.par_state(phase);
+    int state_index = checked_cast<int>(phase_range.index(phase));
```

Comment:
I feel `EnumRange<WeakProcessorPhase>().index(phase)` is better than `phase_range.index(phase)`, since we want to know the "global" index for this phase, not the "local" index within this particular range instance. The two are identical currently. However, if the for-loop is iterating over a subset of all `WeakProcessorPhase`, the two cases will give different results, I believe.

Reply:
I agree that this isn't entirely ideal as-is. If someone were to change
phase_range to be a subset of the full range, then this line would need to
be examined and probably modified. However, that's not going to happen. This
PR is an intermediate step toward eliminating WeakProcessorPhase entirely,
which will address this issue (among others). I have that followup ready for
review once this one is integrated and I've rebased and retested.

Reply:
I see; thanks for the explanation.